app/portainer/views/stacks/edit/stackController.js

```diff
@@ -14,24 +14,28 @@ function ($q, $scope, $state, $transition$, StackService, NodeService, ServiceSe
     Endpoint: null
   };
 
+  $scope.duplicateStack = duplicateStack;
+
   $scope.showEditor = function() {
     $scope.state.showEditorTab = true;
   };
 
-  $scope.migrateStack = function() {
-    ModalService.confirm({
-      title: 'Are you sure?',
-      message: 'This action will deploy a new instance of this stack on the target endpoint, please note that this does NOT relocate the content of any persistent volumes that may be attached to this stack.',
-      buttons: {
-        confirm: {
-          label: 'Migrate',
-          className: 'btn-danger'
+  $scope.migrateStack = function (name, endpointId) {
+    return $q(function (resolve) {
```

Comment:
Any reason for the introduction of the `$q(function (resolve) {})` pattern here?

Reply:
yes, my component is expecting a promise so it can stop the `duplication/migrationInProcess` spinner